### PR TITLE
run update_resource inside a transaction to avoid autosaving relationships through assign_attributes when the record is invalid

### LINKED ISSUE #5430: Associations are persisted before validation of main record
### Expected behavior

When updating a record with changes to its children that make the record fail validation, no changes are saved to the database.

### Actual behavior

The invalid association change is persisted, and then the user gets redirected to a form displaying a validation error.

### How to reproduce

```ruby
require 'bundler/inline'

gemfile(true) do
  source 'https://rubygems.org'

  gem 'activeadmin', ENV['ACTIVE_ADMIN_PATH'] ? {path: ENV['ACTIVE_ADMIN_PATH'], require: false} : {github: 'activeadmin/activeadmin', require: false}
  gem 'rails', '~> 5.1.0'
  gem 'sass-rails'
  gem 'sqlite3', platform: :mri
  gem 'activerecord-jdbcsqlite3-adapter', "~> 51.0", platform: :jruby
end

require 'active_record'

ActiveRecord::Base.establish_connection(adapter: 'sqlite3', database: ':memory:')
ActiveRecord::Base.logger = Logger.new(STDOUT)

ActiveRecord::Schema.define do
  create_table :agents, force: true do |t|
    t.string :full_name
  end

  create_table :agents_clients, force: true do |t|
    t.bigint :agent_id, null: false
    t.bigint :client_id, null: false
  end

  create_table :clients, force: true do |t|
    t.string :name
  end
end

require 'action_controller/railtie'
require 'action_view/railtie'
require 'active_admin'

class TestApp < Rails::Application
  config.root = __dir__
  secrets.secret_token = 'secret_token'
  secrets.secret_key_base = 'secret_key_base'

  config.eager_load = false
  Rails.logger = config.logger = Logger.new($stdout)
end

class ApplicationController < ActionController::Base
  include Rails.application.routes.url_helpers
end

class Client < ActiveRecord::Base
  has_and_belongs_to_many :agents
  validates :agents, presence: true
end

class Agent < ActiveRecord::Base
  has_and_belongs_to_many :clients
end

ActiveAdmin.setup do |config|
  config.authentication_method = false
  config.current_user_method = false
end

Rails.application.initialize!

ActiveAdmin.register Client do
  permit_params agent_ids: []
  
  form do |f|
    f.input :agents, as: :check_boxes
  end
end

Rails.application.routes.draw do
  ActiveAdmin.routes(self)
end

require 'minitest/autorun'
require 'rack/test'

class HabtmTest < ActionDispatch::IntegrationTest
  def test_clients
    Agent.create! full_name: 'John Doe'
    Client.create! name: 'ACME', agents: [Agent.first]

    patch admin_client_url(Client.first), params: {client: {agent_ids: []}}
    assert_response :success
    assert_select "p.inline-errors", "can't be blank"
    assert_equal 1, Client.first.agents.count
  end

  private

  def app
    Rails.application
  end
end


```

This seems to be due to the behavior of `assign_attributes` (as detailed in http://guides.rubyonrails.org/association_basics.html#has-and-belongs-to-many-association-reference). Maybe `update_resource` should check if an `ids` field is present, and if so use `collection.build` for that?

### PR DESCRIPTION
**Pull request by deivid-rodriguez**
A copy of #6202, to not let it be forgotten.

Fixes https://github.com/activeadmin/activeadmin/issues/5430

When submitting a form with invalid attributes, some associations are autosaved via assign_attributes before calling save_resource. As a consequence, when updating a resource, the user sees a validation error and is not aware that the changes to the associations got persisted.

This PR runs the code inside the update_resource method in an ActiveRecord::Transaction so that everything is rolled back when the model is not valid.

### CODE REVIEW

**Review by javierjulio — APPROVED** [review]
Thanks David! Not sure if you wanted to get this in as is or with tests.

**Comment by matgaw** [comment]
@deivid-rodriguez @javierjulio is this planned to be merged at some point? it solves an important issue

**Comment by thibaudgg** [comment]
I also just encountered that issue, this fix is important and should be merged.

**Comment by ggstroligo** [comment]
Is this PR be merged or there is another fix for this issue already on the codebase?

**Comment by deivid-rodriguez** [comment]
Hello, sorry for the radio silence here. I'll try to find some time to re-evaluate this in the following weeks!

**Comment by deivid-rodriguez** [comment]
> Is this PR going to be merged, or is there another fix for this issue already in the codebase?

Are you experiencing this problem? I personally don't know if it has been fixed independently, but if you're experiencing the issue, I think you're very well equipped to verify it!

**Comment by mgrunberg** [comment]
> > Is this PR going to be merged, or is there another fix for this issue already in the codebase?
> 
> Are you experiencing this problem? I personally don't know if it has been fixed independently, but if you're experiencing the issue, I think you're very well equipped to verify it!

@deivid-rodriguez I manage to replicate the original issue in v4 branch adding some models to the template app. This confirms the errors hasn't been fixed.

I want to write a feature test about the change. I don't have time now but should have it in the following days.

**Comment by mgrunberg** [comment]
@deivid-rodriguez I found time to write the spec.

I rebased the PR. The first commit shows the failure and the second one, a cherry-pick of the original code, fixes it.

About the spec. It involved a new model so there are a lot of new files just for this. It may be the wrong call but I considered it cleaner than trying to find a way to mimic the issue scenario with existing models (for example: using `Post` and `Tag`). 

Now we have `Company` with a `has_and_belongs_to`  `Store`. The feature test changes the `Company` to a state where validations fail (`name` can't be blank) and the test expects `stores` changes to be rollbacked.

**Comment by javierjulio** [comment]
@mgrunberg thank you. Did you consider or try doing a controller or unit test instead for this? Cucumber adds significant overhead that makes it hard to follow. Then again I don't know how easy it is to do in a controller or unit test, it would just be ideal if we could I think.

**Comment by mgrunberg** [comment]
> @mgrunberg thank you. Did you consider or try doing a controller or unit test instead for this? Cucumber adds significant overhead that makes it hard to follow. Then again I don't know how easy it is to do in a controller or unit test, it would just be ideal if we could I think.

@javierjulio I didn't consider a controller test. No strong argument here. I simply went for a feature test. It should be possible to add a spec in `spec/unit/resource_controller_spec.rb` (at least I can't think a blocker) but I won't have time to do this for a few days. If that's not a problem I'm ok on give it a try to controller specs.

Just one question, are you fine with the introduction of the new modals?
Because even if I replace the feature spec with a controller one I still need the new model, migrations, etc. The amount of new files won't decrease.

**Comment by javierjulio** [comment]
@mgrunberg no worries, take your time. Thank you. Yes the models, migrations, etc. are ok I just prefer to avoid Cucumber if we can. Although feel free to leave the feature spec as is since it's already written. I just think a unit or controller test would be easier to understand and review if we can. Thank you again.

**Comment by mgrunberg** [comment]
@javierjulio I wrote the controller spec. Let me know what you think

File: spec/unit/resource_controller_spec.rb [hunk 1]
```diff
@@ -124,6 +124,24 @@ def call_after_destroy(obj); end
...
+  describe "resource update" do
+    let(:controller) { Admin::CompaniesController.new }
+
+    around do |example|
+      with_resources_during(example) do
+        ActiveAdmin.register Company
+      end
+    end
+
+    it "should not update habtm associations when validation fails" do
```
Comment (javierjulio) [inline spec/unit/resource_controller_spec.rb:137]:
```suggestion
    it "should not update habtm associations when the resource validation fails" do
```

**Review by javierjulio — APPROVED** [review]
Thanks!

**Comment by Fs00** [comment]
@javierjulio I see this hasn't been backported to 3.x, would you accept a PR that does it? We're hitting the bug fixed by this PR in a production app running AA 3.

**Comment by javierjulio** [comment]
@Fs00 yes. Typically these changes have been backported so I'm not sure if we skipped it for some reason so let's confirm first. @mgrunberg any concern with this change being backported to v3?

**Comment by mgrunberg** [comment]
@javierjulio no problems at all on backporting. I guess it got misplaced. Let me backport this

**Comment by Fs00** [comment]
Many thanks! Looking forward to it 👀🙏

**Comment by mgrunberg** [comment]
@Fs00, @javierjulio backport released as v3.5.2
